homeassistant/helpers/device_registry.py

```diff
@@ -283,19 +316,45 @@ def _async_update_device(
     @callback
     def async_remove_device(self, device_id: str) -> None:
         """Remove a device from the device registry."""
-        del self.devices[device_id]
+        device = self.devices.pop(device_id)
+        self.deleted_devices[device_id] = DeletedDeviceEntry(
+            config_entries=set(device.config_entries),
+            connections=list(device.connections),
+            identifiers=list(device.identifiers),
+            id=device.id,
+        )
         self.hass.bus.async_fire(
             EVENT_DEVICE_REGISTRY_UPDATED, {"action": "remove", "device_id": device_id}
         )
         self.async_schedule_save()
 
+    @callback
+    def async_purge_deleted_device(self, device_id: str) -> None:
+        """Permanently remove a device from the device registry."""
+        del self.deleted_devices[device_id]
+        self.async_schedule_save()
+
+    @callback
+    def async_restore_device(self, device_id: str) -> None:
+        """Restore a deleted device."""
+        deleted_device = self.deleted_devices.pop(device_id)
+        self.devices[device_id] = DeviceEntry(
+            config_entries=deleted_device.config_entries,  # type: ignore
+            connections=deleted_device.connections,  # type: ignore
+            identifiers=deleted_device.identifiers,  # type: ignore
```

Comment:
mypy complains: `Argument "identifiers" to "DeviceEntry" has incompatible type "List[Tuple[str, str]]"; expected "Iterable[_T]"`